src/libPMacc/include/random/distributions/uniform/Uniform_float.hpp

```diff
@@ -34,24 +35,148 @@ namespace distributions
+            return static_cast<result_type>( random ) * value2pow32Inv +
+                ( value2pow32Inv / 2.0f );
+        }
+    };
+
+    /** Returns a random float value uniformly distributed in [0,1)
+     *
+     * Swap the value one to zero (creates a small error in uniform distribution)
+     */
+    template<class T_RNGMethod>
+    class Uniform<
+        uniform::ExcludeOne<float>::SwapOneToZero,
+        T_RNGMethod,
+        void
+    >
+    {
+    public:
+        typedef T_RNGMethod RNGMethod;
+        typedef typename RNGMethod::StateType StateType;
+        typedef float result_type;
+
+        DINLINE result_type
+        operator()(StateType& state) const
+        {
+            const float randomValue =
+                PMacc::random::distributions::Uniform<
+                    uniform::ExcludeZero<result_type>,
+                    RNGMethod
+            >()(state);
+            return randomValue == 1.0f ? 0.0f : randomValue;
+        }
+    };
+
+    /** Returns a random float value uniformly distributed in [0,1)
+     *
+     * Reduce the random range to `2^24`.
+     * Uses a uniform distance of `2^-24` (`epsilon/2`) between each possible
+     * random number.
+     */
+    template<class T_RNGMethod>
+    class Uniform<
+        uniform::ExcludeOne<float>::Use24Bit,
+        T_RNGMethod,
+        void
+    >
+    {
+    public:
         typedef T_RNGMethod RNGMethod;
         typedef typename RNGMethod::StateType StateType;
+        typedef float result_type;
+
+        DINLINE result_type
+        operator()(StateType& state) const
+        {
+            const float value2pow24Inv = 5.9604645e-08f;
+            const float randomValue24Bit = RNGMethod().get32Bits(state) >> 8;
+            return static_cast<result_type>( randomValue24Bit ) * value2pow24Inv;
+        }
+    };
+
+    /** Returns a random float value uniformly distributed in [0,1)
+     *
+     * Loops until a random value inside the defined range is created.
+     * The runtime of this method is not deterministic.
+     */
+    template<class T_RNGMethod>
+    class Uniform<
+        typename uniform::ExcludeOne<float>::Repeat,
+        T_RNGMethod,
+        void
+    >
+    {
     public:
+        typedef T_RNGMethod RNGMethod;
+        typedef typename RNGMethod::StateType StateType;
         typedef float result_type;
 
         DINLINE result_type
-        operator()(StateType& state)
+        operator()(StateType& state) const
         {
-            return _curand_uniform(RNGMethod().get32Bits(state));
+            do
+            {
+                const float randomValue =
+                    PMacc::random::distributions::Uniform<
+                        uniform::ExcludeZero<result_type>,
+                        RNGMethod
+                    >()(state);
+
+                if( randomValue != 1.0f )
+                    return randomValue;
+            }
+            while(true);
```

Comment:
I'd favor `while(randomValue == 1.0f); return randomValue;` as it makes the intent clear. This looks a bit hackish to me.
Generated PTX code is the same however (CUDA 7.5)